### PR TITLE
overflowing-checking for rhs of shift operators

### PR DESCRIPTION
**Pull request by pnkfelix**
overflow-checking for rhs of shift operators

Subtask of #22020 ([RFC 560](https://github.com/rust-lang/rfcs/blob/master/text/0560-integer-overflow.md))

### CODE REVIEW

**Comment by pnkfelix** [comment]
cc @nikomatsakis

File: src/librustc_trans/trans/expr.rs [hunk 1]
```diff
@@ -2477,11 +2507,144 @@ fn with_overflow_check<'a, 'b>(bcx: Block<'a, 'b>, oop: OverflowOp, info: NodeId
...
+    fn build_with_input_check<'blk, 'tcx>(&self, bcx: Block<'blk, 'tcx>,
+                                          info: NodeIdAndSpan,
+                                          lhs_t: Ty<'tcx>, lhs: ValueRef,
+                                          rhs_t: Ty<'tcx>, rhs: ValueRef,
+                                          binop_debug_loc: DebugLoc)
+                                          -> (Block<'blk, 'tcx>, ValueRef)
+    {
+        let to_llty = |r| bcx.ccx().tn().type_to_string(Type::from_ref(r));
+        {
+            let given_llty = type_of::type_of(bcx.ccx(), rhs_t);
```
Comment (pnkfelix) [inline src/librustc_trans/trans/expr.rs]: ah, this whole sub-block is debugging code that is no longer needed; I will remove it.

File: src/librustc_trans/trans/expr.rs [hunk 2]
```diff
@@ -2477,11 +2507,144 @@ fn with_overflow_check<'a, 'b>(bcx: Block<'a, 'b>, oop: OverflowOp, info: NodeId
...
+        //
+        // Note that the mask's value is derived from the LHS type
+        // (since that is where the 32/64 distinction is relevant) but
+        // the mask's type must match the RHS type (since they will
+        // both be fed into a and-binop)
+        let invert_mask = !shift_mask_val(lhs_llty);
+        let invert_mask = C_integral(rhs_llty, invert_mask, true);
+
+        {
+            let inv_llty = val_ty(invert_mask);
```
Comment (pnkfelix) [inline src/librustc_trans/trans/expr.rs]: same here; debugging code, will remove it.

**Comment by pnkfelix** [comment]
One of the tests in this PR uncovered #23551, so don't attempt to merge this until that has a fix.

**Comment by pnkfelix** [comment]
Okay, after discovering that #23551 was a duplicate of #10183 and reading the discussion there further, I have decided to go with @glaebhoerl's [advice](https://github.com/rust-lang/rust/issues/22020#issuecomment-83717008) and ensure that we do not encounter undefined behavior due to the shift's RHS.

**Comment by nikomatsakis** [comment]
r+ -- This looks good to me. I had two questions though so not informing bors just yet (though no action is needed or requested necessarily).

**Comment by pnkfelix** [comment]
@nikomatsakis constant evaluation is not yet addressed.

I don't know whether we have a reliable way to *force* the debug-assertions off, though that sounds like a good thing to have if we do not already have it.  I will investigate.

(Still, we probably can land this and leave those as to-do items.  In particular, we need to deal with const-evaluation for *every* operation; not just shifts, and that still remains to be done.)

**Comment by pnkfelix** [comment]
odd, thought I had done a full test ... or no, maybe I only did `make check` and that does not cover these pretty print tests.

**Comment by pnkfelix** [comment]
(the bug I'm referring to in the commit is #20937 )

**Comment by pnkfelix** [comment]
grar more pretty tests.  (Its hard for me to properly guard against this locally due to #22379 )

**Comment by pnkfelix** [comment]
filed #23623 against the pretty printer; but I think I know how I can work around the problem in the meantime.

**Comment by Manishearth** [comment]
This fails pretty

http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2672/steps/test/logs/stdio

**Comment by pnkfelix** [comment]
hmm, I guess I need to pass `--disable-optimize-tests` to catch cases like this.  (Or manually recreate the effect of that flag locally, to save myself some time...)

**Comment by pnkfelix** [comment]
well, here's hoping the rollup suceeds where this latest merge attempt failed...

**Pull request merged**